src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala

```diff
@@ -0,0 +1,184 @@
+    * database and stop monitoring the cluster.
+    * @param errorDetails cluster error details from Google
+    * @return ShutdownActor
+    */
+  private def handleFailedCluster(errorDetails: ClusterErrorDetails): Future[ClusterMonitorMessage] = {
+    if (shouldRecreateCluster(errorDetails.code, errorDetails.message)) {
+      // Delete the cluster in Google and update the database record to Deleting.
+      // Then shutdown this actor, but register a callback message to the supervisor
+      // telling it to recreate the cluster.
+      logger.info(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is in an error state with $errorDetails. Attempting to delete and recreate...")
+      for {
+        _ <- gdDAO.deleteCluster(cluster.googleProject, cluster.clusterName)
+        _ <- dbRef.inTransaction { _.clusterQuery.deleteCluster(cluster.googleId) }
+      } yield ShutdownActor(Some(ClusterDeleted(cluster, recreate = true)))
+    } else {
+      // Update the database record to Error and shutdown this actor.
+      logger.warn(s"Cluster ${cluster.googleProject}/${cluster.clusterName} is in an error state with $errorDetails'. Unable to recreate cluster.")
+      for {
+        _ <- dbRef.inTransaction(_.clusterQuery.updateClusterStatus(cluster.googleId, ClusterStatus.Error))
+      } yield ShutdownActor()
+    }
+  }
+
+  private def shouldRecreateCluster(code: Int, message: Option[String]): Boolean = {
+    monitorConfig.canRecreateCluster && (code == Code.UNKNOWN.value)
+  }
+
+  /**
+    * Handles a dataproc cluster which has been deleted. We update the status to Deleted in the database,
+    * and shut down this actor.
+    * @return error or ShutdownActor
+    */
+  private def handleDeletedCluster(): Future[ClusterMonitorMessage] = {
+    logger.info(s"Cluster ${cluster.googleProject}/${cluster.clusterName} has been deleted.")
+    dbRef.inTransaction { dataAccess =>
+      dataAccess.clusterQuery.completeDeletion(cluster.googleId, cluster.clusterName)
+    }.map(_ => ShutdownActor())
+  }
+
+  /**
+    * Queries Google for the cluster status and takes appropriate action depending on the result.
+    * @return ClusterMonitorMessage
+    */
+  private def checkCluster: Future[ClusterMonitorMessage] = {
+    val result = for {
+      status <- gdDAO.getClusterStatus(cluster.googleProject, cluster.clusterName)
+      result <- status match {
+        case Unknown | Creating | Updating => Future.successful(NotReadyCluster(status))
+        case Running =>
+          gdDAO.getClusterMasterInstanceIp(cluster.googleProject, cluster.clusterName).map {
+            case Some(ip) => ReadyCluster(ip)
+            case None => NotReadyCluster(ClusterStatus.Running)
+          }
+        case Error =>
+          gdDAO.getClusterErrorDetails(cluster.operationName).map {
+            case Some(errorDetails) => FailedCluster(errorDetails)
+            case None => NotReadyCluster(ClusterStatus.Error)
```

Comment:
so this is different to `NotReadyCluster(Running)` because we don't have the error details, instead of not having the IP details?

Reply:
Correct. Actually the action is the same in either case: we just reschedule another monitor pass.